### PR TITLE
Zombies pushing each other

### PR DESCRIPTION
**Pull request by Coolthulhu**
Kevin suggested something like this not long ago.

A zed who tries to move onto an occupied tile will not cancel immediately, but will try to push aside the current occupant. Chance to push the other zed depends on melee skills and size of the pusher relative to pushee.

The pusher will make a couple of attempts to push the target in random directions. The chance to push is lowered for pushing aside (as opposed to pushing forward).
If the target can't be pushed aside, the pusher will try to "trample" it instead. This only happens if the pusher is much stronger or has a much better roll and will cause both zeds to swap positions and the weaker one to get stunned for a while. Could remove that stun effect - it's more for flavor than gameplay.

If the selected position has a yet another zed, a recursive push attempt will be made, with a boost from the current zed. This is to prevent 3 zeds "hogging" the spots near the player - with enough boosts, even smaller zeds may be able to push the big ones further behind. The recursion is rather limited to keep it from being expensive to compute (max depth 3, each zed can be recursively pushed at most once per turn).

As a result:
- Standing in front of a window and knifing zeds will not work as well
- Angry hordes look like a big blob rather than a rough triangle
- Hiding from brutes behind children will not work

TODO:
- [x] Fix a bug with hordes "boosting" the zeds in front, making them go faster than the fastest zed in the horde

TODO later (not in this PR):
- Make hulks swat zeds aside instead of attacking them
- Make zeds push the player too

### CODE REVIEW

File: src/monmove.cpp [hunk 1]
```diff
@@ -996,6 +997,108 @@ bool monster::move_to( const tripoint &p, bool force )
...
+    }
+
+    // TODO: Generalize this to Creature
+    const int mondex = g->mon_at( p );
+    if( mondex < 0 ) {
+        return false;
+    }
+
+    monster *critter = &g->zombie( mondex );
+    if( critter == nullptr || critter == this || p == pos() ) {
```
Comment (BevapDin) [inline src/monmove.cpp]: Any specific reasons why `critter` is a pointer and not a reference? Note that `critter` can not be NULL, `game::zombie` returns a reference and references can refer to NULL.

Comment (Coolthulhu) [inline src/monmove.cpp]: I want to change it to `critter_at` later, so that it supports pushing the player.

**Comment by DavidKeaton** [comment]
I'd keep the stun effect on trample, or perhaps a prone condition. It is too flavorful mate!

**Comment by kevingranade** [comment]
I think as long as you give the pushed zombie some effect that costs them at least the number of moves it would have taken to make the move themselves, it will all balance out, even if one zombie does get pushed ahead of the pack, they will take several turns to recover.
I'm playtesting this a little, and I think it needs to back off a bit, you only get perhaps a turn or two before the zombies push a friend in through a window after you.

**Comment by DavidKeaton** [comment]
*Cataclysm: Black Friday Sales Ahead*

**Comment by Zireael07** [comment]
Stun or prone on pushed/trampled is a good idea.

**Comment by Coolthulhu** [comment]
Damn, now I see why do they move so fast: while rewriting this function a bunch of times, I forgot to include the move cost for the target in the latest version.

As for the zombies pushing too fast: how about keeping the pushing speed, but making the newly pushed zed get prone if being pushed from a hard to move terrain?
Alternatively I could make the pusher take extra cost for pushing stuff through rough terrain. This cost could be based on push roll difference to make brutes shove regular zeds fast, even through windows.
Possibly both of those.

**Comment by kevingranade** [comment]
Now I have a mental image of zombies laying siege to a fortified house, with hulks throwing child zombies throw windows on the 2nd floor.

**Comment by kevingranade** [comment]
But more seriously, I could see bigger zombies costing themselves less,
though the effect on the shoved zombie should probably be the same.
One thought, prone zombies should probably be immune to shoving, but brutes
and hulks should still fling them. (Also I always planned to have them
fling at players...)

**Comment by Coolthulhu** [comment]
Bigger zeds have better rolls, which lowers the cost of shoving.
If prone zeds were immune to shoving, they'd block shoving totally instead of having zeds walk over them. Or they could be trampled automatically (or just much more easily) - this would move them out of the way pretty quickly.

I didn't touch hulks here - their ATTACKMON currently overrides their PUSH_MON.

**Pull request merged**